Karapace security

juha-aiven opened the pull request:
Add documentation on Karapace Schema Registry and Karapace REST Proxy security.

Note this documentation is for an unreleased feature.

lornajane commented:
Thanks for bringing this! I will admit that I'm confused so far though, do you think we could see some examples of changing the ACLs for both the REST API and the schema registry? And perhaps some information about what happens if access is denied in either of those contexts. Also is this feature so that a user who can't access the schema for a message shouldn't be able to access the message, is that the idea? I think we should explain what this is for as well, if there are one or two clear problems it solves.

juha-aiven commented:
> do you think we could see some examples of changing the ACLs for both the REST API and the schema registry?

I added an example how to add an ACL entry in this PR. This is for Schema Registry only, REST API authz we don't have yet.

> And perhaps some information about what happens if access is denied in either of those contexts. 

Added explanation of the error code.

> Also is this feature so that a user who can't access the schema for a message shouldn't be able to access the message, is that the idea? I think we should explain what this is for as well, if there are one or two clear problems it solves.

I'm not sure if I'm following. The Kafka topic ACL controls access to messages themselves in Kafka. Schema Registry ACL controls the access to the schemas etc stored in Schema Registry. So these are related, but not connected. Both ACLs are managed separately.

ftisiot commented:
Thanks @juha-aiven for the contribution, I'm currently reorganising the content to follow the devportal structure, I'll push a commit soon with the new org and then I'll tag you for more details to the sections I feel missing

ftisiot reviewed (changes requested):
Hi @juha-aiven good work! I reviewed the article and split it following the devportal structure.
Could you please check the various articles if they make sense?

Also it would be good to add:
* a list of Karapace endpoints and the need for read or write options
* A better description of what the `Global:` does... and what options can be associated with it? I didn't fully got it

ftisiot commented:
I believe this is ready to be reviewed, since I've been involved a lot I would hope to have an external reviewer.

I'm a bit concerned that the Karapace endpoints are not well explained, but that might be a joint effort between devportal and karapace docs. But mainly we can tackle it as with an additional PR and have this initial PR go live in devportal as it it since the feature should go live soon

tvainika commented on docs/products/kafka/concepts/schema-registry-authorization.rst:
```diff
@@ -0,0 +1,103 @@
+
+The algorithm to determine if a request has the necessary permissions is the following:
+
+* Iterate over ACL entries in the order
+* Check if the current requesting user matches the ACL entry and the resource in the ACL entry matches the accessed resource
+* If it matches, use that ACL entry. Check whether the operation in the ACL entry matches the accessed operation, based on that grant or deny access
+* If no ACL entry is found, deny access, with HTTP response having status code 401
```
This is not how this has been implemented. The implementation grants access if any of ACL entries matches, and allows access. The order does not matter.

tvainika commented on docs/products/kafka/concepts/schema-registry-authorization.rst:
```diff
@@ -0,0 +1,103 @@
+Karapace schema registry offers a series of endpoints to:
+
+* Get/set the global configurations: ``/config``
+* Get/set schema subjects and associated details: ``/subject``
```
There is also `/compatibility/subjects/<subject:path>/versions/<version:path>`, `/schemas/ids/<schema_id:path>/versions`, and `/schemas/ids/<schema_id:path>` endpoints related to authorization.

tvainika commented on docs/products/kafka/concepts/schema-registry-authorization.rst:
```diff
@@ -0,0 +1,103 @@
+    - Read access for ``user_1`` to the subject ``s1``. When filtering results in list endpoints, only return the results related to subjects ``s1``.
+  * - ``user_1``
+    - ``schema_registry_write``
+    - ``Subject: s1``
```
```suggestion
    - ``Subject:s1``
```

tvainika commented on docs/products/kafka/concepts/schema-registry-authorization.rst:
```diff
@@ -0,0 +1,103 @@
+    - Read access for ``user_1`` to the global compatibility configuration
+  * - ``user_1``
+    - ``schema_registry_read``
+    - ``Subject: s1``
```
```suggestion
    - ``Subject:s1``
```

tvainika commented on docs/products/kafka/concepts/schema-registry-authorization.rst:
```diff
@@ -0,0 +1,103 @@
+    - Write access for ``user_1`` to the subject ``s1``. Write implies read access, see the above row.
+  * - ``user_readonly*``
+    - ``schema_registry_read``
+    - ``Subject: s*``
```
```suggestion
    - ``Subject:s*``
```

tvainika commented on docs/products/kafka/concepts/schema-registry-authorization.rst:
```diff
@@ -0,0 +1,103 @@
+    - Read access for users with prefix ``user_readonly`` to the subjects with prefix ``s``. When filtering results, only return entities for subjects with prefix ``s``.
+  * - ``user_write*``
+    - ``schema_registry_write``
+    - ``Subject: s*``
```
```suggestion
    - ``Subject:s*``
```

tvainika commented on docs/products/kafka/howto/enable-schema-registry-authorization.rst:
```diff
@@ -0,0 +1,26 @@
+
+Enable Karapace schema registry authorization via Aiven CLI
+-----------------------------------------------------------
+
```
Should we mention this requires newest aiven CLI too?

tvainika commented on docs/tools/cli/service/schema-registry-acl.rst:
```diff
@@ -0,0 +1,85 @@
+::
+
+  avn service schema-registry-acl-add kafka-doc \
+    --username userAB*                          \
```
How about commands here and all over, if this is copy-pasted to local shell and cwd contains file named userABC, this will cause unexpected behavior? How about quoting `'userAB*'`?

tvainika commented on docs/tools/cli/service/schema-registry-acl.rst:
```diff
@@ -0,0 +1,85 @@
+
+::
+
+  avn service schema-registry-acl-list kafka-doc
```
How about adding example output here?

tvainika reviewed (changes requested):
Some misunderstandings found, and maybe some other useful suggestions

TibsAtWork reviewed (approved):
* Added a few ® marks
* Split apart references to "Aiven for Apache Kafka® Karapace" to make it clearer that Karapace is a separate thing.
* Made some minor changes to the english

I believe I've not changed any *meaning* of anything, and thus this can be approved.
